backtesting/backtesting.py

```diff
@@ -454,128 +653,248 @@ def __init__(self, *, data, cash, commission, margin, trade_on_close, length):
+
+            # Check if stop condition was hit
+            stop_price = order.stop
+            if stop_price:
+                is_stop_hit = ((high > stop_price) if order.is_long else (low < stop_price))
+                if not is_stop_hit:
+                    continue
+
+                # > When the stop price is reached, a stop order becomes a market/limit order.
+                # https://www.sec.gov/fast-answers/answersstopordhtm.html
+                order._replace(stop_price=None)
+
+            # Determine purchase price.
+            # Check if limit order can be filled.
+            if order.limit:
+                is_limit_hit = low < order.limit if order.is_long else high > order.limit
+                # When stop and limit are hit within the same bar, we pessimistically
+                # assume limit was hit before the stop (i.e. "before it counts")
+                is_limit_hit_before_stop = (is_limit_hit and
+                                            (order.limit < (stop_price or -np.inf)
+                                             if order.is_long
+                                             else order.limit > (stop_price or np.inf)))
+                if not is_limit_hit or is_limit_hit_before_stop:
+                    continue
+
+                # stop_price, if set, was hit within this bar
+                price = (min(open, order.limit, stop_price or np.inf)
+                         if order.is_long else
+                         max(open, order.limit, stop_price or -np.inf))
+            else:
+                # Market-if-touched / market order
+                price = prev_close if self._trade_on_close else open
+                price = (max(price, stop_price or -np.inf)
+                         if order.is_long else
+                         min(price, stop_price or np.inf))
+
+            # Determine entry/exit bar index
+            is_market_order = not order.limit and not stop_price
+            time_index = (self._i - 1) if is_market_order and self._trade_on_close else self._i
+
+            # If order is a SL/TP order, it should close an existing trade it was contingent upon
+            if order.parent_trade:
+                trade = order.parent_trade
+                # If this trade isn't already closed (e.g. on multiple `trade.close(.5)` calls)
+                if trade in self.trades:
+                    self._reduce_trade(trade, price, order.size, time_index)
+                    assert order.size != -trade.size or trade not in self.trades
+                if order in (trade._sl_order,
+                             trade._tp_order):
+                    assert order.size == -trade.size
+                    assert order not in self.orders  # Removed when trade was closed
+                else:
+                    # It's a trade.close() order, now done
+                    assert abs(trade.size) >= abs(order.size) >= 1
+                    self.orders.remove(order)
+                continue
```

Comment:
> https://github.com/kernc/backtesting.py/pull/47#issuecomment-635372131

@sdmovie Explicit closing orders (SL/TP orders and orders made through `self.trades[i].close(portion)` or `self.position.close(portion)`) are already exempt from margin computation.

Reply:
Great! that position close already fullfill my use case. Not necessary to struggle on buy/sell margin calculation - for me

Reply:
how does resample work to comply various open hours (start time unalign and segmented hours)? 
US stock 9:30-16:00 - should be 9:30-10:30,10:30-11:30,...15:00-16:00
US future 17:00 -16:15 should be 17:00-18:00,... 16:00-16:15
HK stock: 9:30-12:00,13:00-16:00  -- should be 9:30-10:30,10:30-11:30,11:30-12:00,13:00-14:00 .. 15:00-16:00

I have s draft proposal: